crates/engine/src/rocksdb_engine/mod.rs

```diff
@@ -200,51 +201,71 @@ impl StorageEngine for RocksEngine {
+                        let cf = self
+                            .inner
+                            .cf_handle(table)
+                            .ok_or(EngineError::TableNotFound((*table).to_owned()))?;
+                        transaction.delete_cf(&cf, key)?;
+                    }
+                    WriteOperation::DeleteRange { table, from, to } => {
+                        let cf = self
+                            .inner
+                            .cf_handle(table)
+                            .ok_or(EngineError::TableNotFound((*table).to_owned()))?;
+                        let mode = IteratorMode::From(from, Direction::Forward);
+                        let kvs: Vec<_> = transaction
+                            .iterator_cf(&cf, mode)
+                            .take_while(|res| res.as_ref().is_ok_and(|(key, _)| key.as_ref() < *to))
+                            .collect::<Result<Vec<_>, _>>()?;
+                        for (key, _) in kvs {
+                            transaction.delete_cf(&cf, key)?;
+                        }
+                    }
                 }
-                WriteOperation::Delete { table, key } => {
-                    let cf = self
-                        .inner
-                        .cf_handle(table)
-                        .ok_or(EngineError::TableNotFound(table.to_owned()))?;
-                    transaction.delete_cf(&cf, key)?;
+            }
+            match transaction.commit() {
+                Ok(_) => {
+                    _ = self
+                    .size
+                    .fetch_add(size.numeric_cast(), std::sync::atomic::Ordering::Relaxed);
+                    return Ok(());
                 }
-                WriteOperation::DeleteRange { table, from, to } => {
-                    let cf = self
-                        .inner
-                        .cf_handle(table)
-                        .ok_or(EngineError::TableNotFound(table.to_owned()))?;
-                    let mode = IteratorMode::From(from, Direction::Forward);
-                    let kvs: Vec<_> = transaction
-                        .iterator_cf(&cf, mode)
-                        .take_while(|res| res.as_ref().is_ok_and(|kv| kv.0.as_ref() < to))
-                        .collect::<Result<Vec<_>, _>>()?;
-                    for (key, _) in kvs {
-                        transaction.delete_cf(&cf, key)?;
+                Err(err) if matches!(err.kind(), RocksErrorKind::Busy | RocksErrorKind::TryAgain) => {
+                    if retry_count > max_retry_count {
+                        warn!("Oops, txn commit retry count reach the max_retry_count: {max_retry_count}");
+                        return Err(EngineError::UnderlyingError(err.to_string()));
                     }
+                    warn!("Rocksdb txn commit failed, retrying after {retry_interval}ms");
+                    std::thread::sleep(std::time::Duration::from_millis(retry_interval));
```

Comment:
Using a thread::sleep will block a tokio worker thread, I suggest to retry immediately as the conflict situation is rare.

Reply:
Retrying immediately when a conflict occurs does nothing to help; instead, it aggravates the conflict.